Vk/add app setup wizard skeleton

vovakulikov opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/52123

This PR adds the initial UI and the basic layout changes for the new App specific setup flow. 

## Todo 
- Polish local repositories widget step 
- Check the sign-in flow for the first setup step 

## Test plan
- Check that you can see the setup flow at the very beginning when the app is starting 
- Check that you can sing into the .com and go back to the next setup step 
- Check that you can add local repositories 
- Check that you can finish the flow and see the home page after 
- Check that once you finished the flow you don't see it again next run

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sourcegraph-buildkite commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.03% (+0.80 kb) | 0.10% (+15.60 kb) 🔺 | 0.12% (+14.80 kb) 🔺 | -0.26% (-2) 🔽 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/add-app-setup-wizard-skeleton/compare-report.html;tab=live_object) for a full comparison between the commits a508fff and 0c955f4 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

vovakulikov commented:
@limitedmage @marekweb I think this is ready for your review. I didn't check the deep link integration because it doesn't work in the dev mode, @marekweb, how did you test it during your work on this? Should I build a binary with the app and run the binary? 

Also, if anyone knows how to do it, I added a few links to the doc in this PR, but I didn't use the `/help` route since it doesn't redirect you to the doc site properly. I used the full URL, but our link checker keeps complaining about it. Is there a way to make an exception for these links?

limitedmage commented on client/web/src/LegacyLayout.tsx:
```diff
@@ -154,10 +159,30 @@ export const LegacyLayout: FC<LegacyLayoutProps> = props => {
     // setup wizard state, since we don't have a good solution for this at the
     // moment, we use mutable window.context object here.
     // TODO remove window.context and use injected context store/props
```
It doesn't look like `window.context` is being used anymore, is this a stale comment?

vovakulikov replied:
`needsRepositoryConfiguration` is a variable alias for the window.context.needsRepositoryConfiguration I will inline the `window.context` here to make it more explicit

limitedmage commented on client/web/src/LegacyLayout.tsx:
```diff
@@ -154,10 +159,30 @@ export const LegacyLayout: FC<LegacyLayoutProps> = props => {
+    // Redirect to the app setup pages if it's App, we haven't seen setup before,
+    // and it's not already a setup page
+    if (props.isSourcegraphApp && !wasAppSetupFinished && !isAppSetupPage) {
+        return <Navigate to={EnterprisePageRoutes.AppSetup} replace={true} />
```
Note for me: We should not navigate to the setup wizard if this is the Cody standalone page. I'll work on this once we can detect if Cody can be enabled. The Cody page can show an error message if the setup has not been completed.

limitedmage commented on client/web/src/LegacyLayout.tsx:
```diff
@@ -154,10 +159,30 @@ export const LegacyLayout: FC<LegacyLayoutProps> = props => {
+        return <Navigate to={EnterprisePageRoutes.AppSetup} replace={true} />
+    }
+
+    // Some route by their design requires rendering on the blank page
+    // without some general and standard UI that Layout component renders
+    // by default. If route has handle: { fullPage: true } we render just
+    // routes and its container block without rendering global nav, notepad
+    // and other standard UI elements.
```
minor wording suggestions

```suggestion
    // Some routes by their design require rendering on a blank page
    // without the UI chrome that Layout component renders by default.
    // If route has handle: { fullPage: true } we render just the route content
    // and its container block without rendering global nav, notepad
    // and other standard UI chrome elements.
```

limitedmage commented on client/web/src/enterprise/app/setup/AppSetupSteps.tsx:
```diff
@@ -0,0 +1,194 @@
+
+            <div className={styles.actions}>
+                <div className={styles.actionsCard}>
+                    <H3>You’ll need a Sourcegraph.com account in order to connect Cody.</H3>
```
a11y: This should be an `h2` to have consistent ordering of headings (heading levels should not be skipped)

limitedmage commented on client/web/src/enterprise/app/setup/AppSetupSteps.tsx:
```diff
@@ -0,0 +1,194 @@
+    )
+}
+
+const SourcegraphLogo: FC = () => (
```
We already have the Sourcegraph logo in the bundle, can we reuse it? Or is this a different version of the logo?

limitedmage commented on client/web/src/enterprise/app/setup/AppSetupSteps.tsx:
I would suggest splitting each of these into their own file inside a `steps` subfolder, for ease of browsing the code.

vovakulikov replied:
Also agreed, I put them in one file because, at the moment they don't have any big logic, mostly UI and all three steps share layout primitives, a good solution here would be extracting the shared part but it also will create its own module file. 

I think I will wait for the follow-up PR and see what kind of other layout/logic changes all steps will share (I see that there is new layouts in Figma) and them split the components.

slimsag commented:
> I didn't check the deep link integration because it doesn't work in the dev mode, @marekweb, how did you test it during your work on this? Should I build a binary with the app and run the binary?

You can make a release build using:

```
CODESIGNING=0 ./enterprise/dev/app/build-release.sh
```

You can clear the webkit cache directory / localstorage using this (so the setup wizard shows up again):

```
rm -rf ~/.sourcegraph-psql ~/Library/Application\ Support/com.sourcegraph.* ~/Library/Caches/com.sourcegraph.* ~/Library/WebKit/com.sourcegraph.*
```

slimsag commented:
It looks like the auth flow doesn't work right now, not sure why. `Connect to Sourcegraph.com` brings me to the access token page, but once you continue the app just stays on the same screen / doesn't transition to the local repos screen:

<img width="1136" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/3173176/8c9760d9-7c76-423f-8bf4-b1d810299dfa">

I can see that Sourcegraph.com's redirect brought me to:

http://localhost:3080/app/auth/callback?code=sgp_2218aa345e745ab6d9dcee4e22db3f874a5fb77a&destination=%2Fapp-setup%2Flocal-repositories

But there are no page changes after that. No errors in the console either:

<img width="1037" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/3173176/e3b4fcce-3b7d-4e52-985b-a63326e518e6">

slimsag commented:
I dug into the issue deeper, I'm not sure how to fix it but I at least understand what's going on:

1. Sourcegraph.com brings you to e.g. sourcegraph://app/auth/callback?code=$TOKEN&destination=%2Fapp-setup%2Flocal-repositories
2. The app correctly redirects you to http://localhost:3080/app/auth/callback?code=$TOKEN&destination=%2Fapp-setup%2Flocal-repositories
3. The logic in `LegacyLayout.tsx` (I believe) forces the page to be "the last step" which is http://localhost:3080/app-setup/welcome (this is the URL you end up on, despite no request in the network tab) - which prevents `AppAuthCallbackPage.tsx` from rendering, and as a result means the site config is not updated with the access token so the setup wizard doesn't know to go to the next screen.

vovakulikov commented on client/shared/src/settings/temporary/TemporarySettings.ts:
```diff
@@ -73,6 +73,7 @@ export interface TemporarySettingsSchema {
     'search.notebooks.minSavedPerView': number
     'repo.commitPage.diffMode': DiffMode
     'setup.activeStepId': string
+    'app-setup.activeStepId': string
```
@limitedmage yeah, having `setupCompleted` makes sense to me, I actually did this at the beginning, but then it turned out that we have to work with variables about finish state in a sync way because routing depends on this variable and we should have this in sync overwise we will have flashes and unpredictable redirection. 

This is the main reason why I used local storage variable (because it's sync)

limitedmage replied:
That's fine, let's keep it in localStorage if it simplifies things.

slimsag commented:
@vovakulikov I think we can add an exclusion for the /help linter here:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/sg/linters/client.go?L23

```diff
 return diff.IterateHunks(func(file string, hunk repo.DiffHunk) error {
+    if file == "client/web/src/enterprise/app/setup/AppSetupSteps.tsx" {
+        continue
+    }
```

Want me to push a commit for it? Just don't want to conflict :)

Also I'm trying the auth flow again now

vovakulikov commented:
@slimsag I fixed all problems by Julianna's comments and also fixed the auth flow also found a few layout problems in the production build. I added Sourcegraph App directory as an exclusion for the doc links check.

slimsag commented:
Amazing!
